homeassistant/components/mysensors.py

```diff
@@ -0,0 +1,219 @@
+        return False
+
+    version = config[DOMAIN].get(CONF_VERSION, '1.4')
+
+    global CONST
+    if version == '1.4':
+        import mysensors.const_14 as const
+        CONST = const
+    elif version == '1.5':
+        import mysensors.const_15 as const
+        CONST = const
+    else:
+        import mysensors.const_14 as const
+        CONST = const
+
+    # Just assume celcius means that the user wants metric for now.
+    # It may make more sense to make this a global config option in the future.
+    global IS_METRIC
+    IS_METRIC = (hass.config.temperature_unit == TEMP_CELCIUS)
+
+    def callback_generator(port, devices):
+        """Return a new callback function. Run once per gateway setup."""
+        def node_update(update_type, nid):
+            """Callback for node updates from the MySensors gateway."""
+            _LOGGER.info('update %s: node %s', update_type, nid)
+
+            hass.bus.fire(EVENT_MYSENSORS_NODE_UPDATE, {
+                ATTR_PORT: port,
+                ATTR_DEVICES: devices,
+                ATTR_UPDATE_TYPE: update_type,
+                ATTR_NODE_ID: nid
+            })
+            return
+        return node_update
+
+    def setup_gateway(port, persistence, persistence_file):
+        """Return gateway after setup of the gateway."""
+        devices = {}    # keep track of devices added to HA
+        gateway = mysensors.SerialGateway(port,
+                                          persistence=persistence,
+                                          persistence_file=persistence_file,
+                                          protocol_version=version)
+        gateway.event_callback = callback_generator(port, devices)
+        gateway.metric = IS_METRIC
+        gateway.debug = config[DOMAIN].get(CONF_DEBUG, False)
+        gateway.start()
+
+        def persistence_update(event):
+            """Callback to trigger update from persistence file."""
+            for _ in range(2):
+                for nid in gateway.sensors:
+                    gateway.event_callback('persistence', nid)
+
+        if persistence:
+            hass.bus.listen_once(
+                EVENT_HOMEASSISTANT_START, persistence_update)
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP,
+                             lambda event: gateway.stop())
+        return gateway
+
+    port = config[DOMAIN].get(CONF_PORT)
+    persistence_file = config[DOMAIN].get(
+        CONF_PERSISTENCE_FILE, hass.config.path('mysensors.pickle'))
+
+    if isinstance(port, str):
+        port = [port]
+    if isinstance(persistence_file, str):
+        persistence_file = [persistence_file]
+
+    # Setup all ports from config
+    global GATEWAYS
+    GATEWAYS = {}
+    for index, port_item in enumerate(port):
+        persistence = config[DOMAIN].get(CONF_PERSISTENCE, True)
+        try:
+            persistence_f_item = persistence_file[index]
+        except IndexError:
+            _LOGGER.exception(
```

Comment:
If I disable persistence, this warning will still be given. Seems a bit too verbose. Also use `_LOGGER.error` so you don't print the full stacktrace of the exception